### DISCUSSION TITLE
Issues with dJdbeta Calculation: writesensMap Does Not Work Correctly in Parallel

### DISCUSSION THREAD

**Discussion by lnhang0618** [mdolab/dafoam#630] [General]
Hello Professor He,

I am currently performing Field Inversion on the CBFS test case. I have not used cyclic boundaries.However, I am encountering an issue where dJdbeta becomes confused or "messed up" during parallel computations. I have confirmed that I am retrieving dJdbeta using writesensMap and reconstractPar. Could you advise on how I might investigate the cause of this issue?
[CBFS.zip](https://github.com/mdolab/dafoam/files/15024207/CBFS.zip)

![image](https://github.com/mdolab/dafoam/assets/77599661/fe39e6fd-cd97-4ac1-aef0-86c50b1f198d)
In parallel calculations, the derivatives from the first half are getting mixed up or incorrect, which seems to be a malfunction of the writesensMap function.
Additionally, the derivatives for x in the range 10-15 are computed as zero, which occurs regardless of whether the calculations are performed in parallel or not. While small derivative values are expected, the result being exactly zero is unusual and concerning.

**Comment by lnhang0618** [comment]
The grid for this test case, sourced from 'A curated dataset for data-driven turbulence modelling,' differs from the regular grids I've used previously, showing irregularities such as different grid counts at the inlet and outlet (134/145). I suspect these irregularities might affect the adjoint solver's performance.

**Comment by friedenhe** [comment]
Have you tried to compute the dJdBeta in serial? Do they look the same? Also, from the DAFoam's regression test runTest_DASimpleFoamField.py, the field derivs are consistent between serial and parallel runs. So, I am not sure why your case did not work..
